### PR TITLE
feat(db): implement `RefundInterface` for `MockDb`

### LINKED ISSUE #1276: [FEATURE] Implement `RefundInterface` for `MockDb`
Spin out from #172. Please refer to that issue for more information.

### PR DESCRIPTION
**Pull request by younes-io**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
The main motivation is to have MockDb stubs, help to void mocking, and invocation of external database api's.
For more information check https://github.com/juspay/hyperswitch/issues/172

Fixes #1276 

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/db/refund.rs [hunk 1]
```diff
@@ -719,23 +745,39 @@ impl RefundInterface for MockDb {
...
-        Err(errors::StorageError::MockDbError)?
+        let refunds = self.refunds.lock().await;
+
+        Ok(refunds
+            .iter()
+            .take_while(|refund| {
+                refund.merchant_id == merchant_id
+                    && refund_details.limit.is_some()
+                    && refund_details.limit == Some(limit)
+            })
```
Comment (Narayanbhat166) [inline crates/router/src/db/refund.rs:779]: same here, use `filter()` instead. Also use the [take()](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take) method to apply the limit.

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/router/src/db/refund.rs (hunk 1 above)
Comment (younes-io) [reply crates/router/src/db/refund.rs:779]: `take` has an argument of type `usize` whileas `limit` is of type `i64`
what would you suggest ?

File: crates/router/src/db/refund.rs [hunk 2]
```diff
@@ -670,12 +679,29 @@ impl RefundInterface for MockDb {
...
+                *r = refund_updated.clone();
+                refund_updated
+            }) {
+            Some(refund_updated) => Ok(refund_updated),
+            None => {
+                return Err(errors::StorageError::ValueNotFound(
+                    "cannot find refund to update".to_string(),
+                )
+                .into())
+            }
```
Comment (SanchithHegde) [inline crates/router/src/db/refund.rs:703]: You can use [`ok_or_else()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else) here.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

Thanks for the PR, @younes-io!

File: crates/router/src/db/refund.rs [hunk 3]
```diff
@@ -719,23 +745,34 @@ impl RefundInterface for MockDb {
...
+        limit: i64,
     ) -> CustomResult<Vec<storage_models::refund::Refund>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        let refunds = self.refunds.lock().await;
+
+        Ok(refunds
+            .iter()
+            .filter(|refund| refund.merchant_id == merchant_id)
+            .take(limit as usize)
```
Comment (Narayanbhat166) [inline crates/router/src/db/refund.rs:774]: This could panic, use `try_from` here,

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Comment by Narayanbhat166** [comment]
If possible try to handle the unwrap by raising the error. Apart from that Looks good to me. 

Thanks for the Contribution @younes-io!

File: crates/router/src/db/refund.rs [hunk 4]
```diff
@@ -719,23 +740,34 @@ impl RefundInterface for MockDb {
...
+        limit: i64,
     ) -> CustomResult<Vec<storage_models::refund::Refund>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        let refunds = self.refunds.lock().await;
+
+        Ok(refunds
+            .iter()
+            .filter(|refund| refund.merchant_id == merchant_id)
+            .take(usize::try_from(limit).unwrap())
```
Comment (SanchithHegde) [inline crates/router/src/db/refund.rs:769]:
```suggestion
            .take(usize::try_from(limit).unwrap_or(usize::MAX))
```

**Review by SanchithHegde — COMMENTED** [review]
One minor change needed. Other than that, looks good to me!

Edit: Also, please address the failing CI checks.

**Comment by younes-io** [comment]
> One minor change needed. Other than that, looks good to me!
> 
> Edit: Also, please address the failing CI checks.

I ran `cargo clippy --all` locally and I didn't get any errors. I can't see the CLI errors

**Comment by SanchithHegde** [comment]
> I ran `cargo clippy --all` locally and I didn't get any errors. I can't see the CLI errors

The one warning that was thrown previously should be fixed by your latest commit.

**Review by Narayanbhat166 — APPROVED** [review]

**Comment by SanchithHegde** [comment]
Thanks again for the PR, @younes-io!

**Pull request merged**